receiver/googlecloudspannerreceiver/internal/statsreader/intervalstatsreader_test.go

```diff
@@ -0,0 +1,160 @@
+}
+
+func TestNewIntervalStatsReaderWithMaxRowsLimit(t *testing.T) {
+	databaseId := datasource.NewDatabaseId(projectId, instanceId, databaseName)
+	ctx := context.Background()
+	client, _ := spanner.NewClient(ctx, "")
+
+	database := datasource.NewDatabaseFromClient(client, databaseId)
+
+	metricsMetadata := &metadata.MetricsMetadata{
+		Name: "name",
+	}
+
+	logger := zap.NewNop()
+
+	reader := newIntervalStatsReaderWithMaxRowsLimit(logger, database, metricsMetadata, true, topMetricsQueryMaxRows)
+
+	assert.Equal(t, database, reader.database)
+	assert.Equal(t, logger, reader.logger)
+	assert.Equal(t, metricsMetadata, reader.metricsMetadata)
+	assert.Equal(t, topMetricsQueryMaxRows, reader.topMetricsQueryMaxRows)
+	assert.True(t, reader.backFillEnabled)
+}
+
+func TestNewIntervalStatsReader(t *testing.T) {
+	databaseId := datasource.NewDatabaseId(projectId, instanceId, databaseName)
+	ctx := context.Background()
+	client, _ := spanner.NewClient(ctx, "")
+
+	database := datasource.NewDatabaseFromClient(client, databaseId)
+
+	metricsMetadata := &metadata.MetricsMetadata{
+		Name: "name",
+	}
+
+	logger := zap.NewNop()
+
+	reader := newIntervalStatsReader(logger, database, metricsMetadata, true)
+
+	assert.Equal(t, database, reader.database)
+	assert.Equal(t, logger, reader.logger)
+	assert.Equal(t, metricsMetadata, reader.metricsMetadata)
+	assert.True(t, reader.backFillEnabled)
+	assert.Equal(t, 0, reader.topMetricsQueryMaxRows)
+}
+
+func TestPullTimestamps(t *testing.T) {
+	nowAtStartOfMinute := nowAtStartOfMinute()
+	backFillIntervalAgo := nowAtStartOfMinute.Add(-1 * backFillIntervalDuration)
+	backFillIntervalAgoWthSomeSeconds := backFillIntervalAgo.Add(-15 * time.Second)
+
+	testCases := map[string]struct {
+		lastPullTimestamp  time.Time
+		backFillEnabled    bool
+		amountOfTimestamps int
+	}{
+		"Zero last pull timestamp without back filling":                                                       {time.Time{}, false, 1},
+		"Zero last pull timestamp with back filling":                                                          {time.Time{}, true, int(backFillIntervalDuration.Minutes())},
+		"Last pull timestamp now at start of minute back filling does not matter":                             {nowAtStartOfMinute, false, 1},
+		"Last pull timestamp back fill interval ago of minute back filling does not matter":                   {backFillIntervalAgo, false, int(backFillIntervalDuration.Minutes())},
+		"Last pull timestamp back fill interval ago with some seconds of minute back filling does not matter": {backFillIntervalAgoWthSomeSeconds, false, int(backFillIntervalDuration.Minutes()) + 1},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			timestamps := pullTimestamps(testCase.lastPullTimestamp, testCase.backFillEnabled)
+
+			assert.Equal(t, testCase.amountOfTimestamps, len(timestamps))
+		})
+	}
+}
+
+func TestPullTimestampsWithMinuteDifference(t *testing.T) {
```

Comment:
For this test to validate upper bound and lower bound, you may need to control nowStartOfTime() function to return the time you would like. 

Other option is to control time.Now().UTC() function with a wrapper.

Reply:
It is interesting: what is wrong with current approach and why test can't rely on existing approach?

Reply:
I think I'll refactor this a little bit - already have some ideas..